### PR TITLE
[JENKINS-66986] Allow non kubernetes nodes to be provisioned inside the container clause

### PR DESCRIPTION
**Pull request by presPetkov**
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

**Description:**
When a non Kubernetes node is provisioned inside the container clause the steps executions fail because it is trying to cast a DumbSlave to KubernetesSlave - https://github.com/jenkinsci/kubernetes-plugin/blob/master/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java#L255. I added a check and if the node is not an instance of KubernetesSlave it just returns the original launcher.

**JIRA Link:** https://issues.jenkins.io/browse/JENKINS-66986

_No upstream or downstream changes._

**Before change**

image

image

image

image

**After change**

image

image

image

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 1]
```diff
@@ -246,6 +246,12 @@ public void setNodeContext(KubernetesNodeContext nodeContext) {
 
     @Override
     public Launcher decorate(final Launcher launcher, final Node node) {
+
+        //Allows other nodes to be provisioned inside the container clause
+        //If the node is not a KubernetesSlave return the original launcher
+        if(node != null && !(node instanceof KubernetesSlave)) {
+           return new Launcher.DecoratedLauncher(launcher) {};
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java:253]:
```suggestion
           return launcher;
```

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecoratorTest.java [hunk 2]
```diff
@@ -424,6 +427,23 @@ public void testContainerExecOnCustomWorkingDirWithComputeEnvVars() throws Excep
...
+     * @throws Exception
+     */
+    @Test
+    @Issue("JENKINS-66986")
+    public void testRunningANonKubernetesNodeInsideContainerClause() throws Exception {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        DummyLauncher dummyLauncher = new DummyLauncher(new StreamTaskListener(new TeeOutputStream(out, System.out)));
+        Launcher launcher = decorator.decorate(dummyLauncher, dumbAgent);
+        //Assert a DecoratedLauncher was returned successfully
+        assertTrue(launcher != null && launcher instanceof DecoratedLauncher);
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecoratorTest.java:444]:
```suggestion
        assertEquals(dummyLauncher, launcher);
```

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecoratorTest.java [hunk 3]
```diff
@@ -424,6 +427,23 @@ public void testContainerExecOnCustomWorkingDirWithComputeEnvVars() throws Excep
...
+     *
+     * @throws Exception
+     */
+    @Test
+    @Issue("JENKINS-66986")
+    public void testRunningANonKubernetesNodeInsideContainerClause() throws Exception {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        DummyLauncher dummyLauncher = new DummyLauncher(new StreamTaskListener(new TeeOutputStream(out, System.out)));
+        Launcher launcher = decorator.decorate(dummyLauncher, dumbAgent);
+        //Assert a DecoratedLauncher was returned successfully
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecoratorTest.java:443]:
```suggestion
```

**Review by Vlatombe — COMMENTED** [review]
Looks good overall. Can be simplified though.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java (hunk 1 above)
Comment (presPetkov) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java:253]: Ok, thanks. I will try these changes.

**Comment by presPetkov** [comment]
**Test after changes**

image

image

image

**Review by Vlatombe — APPROVED** [review]

**Pull request merged**